### PR TITLE
Auto compile when static cache

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Add automatic compile for `static` cache. 
This can be tested with:
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
import time
device = "cuda"
model = AutoModelForCausalLM.from_pretrained("meta-llama/Meta-Llama-3-8B", torch_dtype=torch.float16).to(device)
tokenizer = AutoTokenizer.from_pretrained("meta-llama/Meta-Llama-3-8B")

sequence = "Hey what's the plan"

inputs = tokenizer.encode(sequence, return_tensors='pt').to(device)
model.generation_config.temperature = 1.0
model.generation_config.top_p = 1.0

t0 = time.time()
out = model.generate(inputs, do_sample=False, max_new_tokens=500)
out = tokenizer.batch_decode(out, skip_special_tokens=True)[0]
dt = time.time() - t0
print(f'dt: {dt}', out)

t0 = time.time()
out = model.generate(inputs, do_sample=False, max_new_tokens=500, cache_implementation="static")
out = tokenizer.batch_decode(out, skip_special_tokens=True)[0]
dt = time.time() - t0
print(f'dt: {dt}', out)

t0 = time.time()
out = model.generate(inputs, do_sample=False, max_new_tokens=500, cache_implementation="static")
out = tokenizer.batch_decode(out, skip_special_tokens=True)[0]
dt = time.time() - t0
print(f'dt: {dt}', out)
```
which give 15sec for dynamic, 30 for the first generate, then 4seconds for the next one

### CODE REVIEW

**Review by Cyrilvallez — COMMENTED** [review]
On principle this LGTM, however in practice I am not experiencing any speedup (but aggravated performance) until the number of new tokens is quite high, of the order of ~2500/3000 (quick test with Llama 3.1 8B)
Note sure if it comes only from compilation time and warmup, or some graph breaks somewhere
Did you try to compare performances a bit?

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -3222,6 +3222,14 @@ def _sample(
         unfinished_sequences = torch.ones(batch_size, dtype=torch.long, device=input_ids.device)
         model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)
 
+        def model_forward(model, *args, **kwargs):
+            return model.forward(*args, **kwargs)
+
+        if isinstance(model_kwargs.get("past_key_values"), StaticCache):
+            if self.device == "cuda":
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/generation/utils.py:3229]: Which device are you using?

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -3222,6 +3223,16 @@ def _sample(
         unfinished_sequences = torch.ones(batch_size, dtype=torch.long, device=input_ids.device)
         model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)
 
+        def model_forward(model, *args, **kwargs):
+            return model.forward(*args, **kwargs)
```
Comment (BenjaminBossan) [inline src/transformers/generation/utils.py:3227]: @ArthurZucker This PR breaks some tests on PEFT :(

I checked why that is exactly, and I found that it has nothing to do with the compilation. The sole reason is that on 2nd iteration, we use this function, which effectively calls:

`self.forward(**model_inputs, return_dict=True)`

whereas on the first iteration (and before the PR), we would call:

`self(**model_inputs, return_dict=True)`

Is there any specific reason why `forward` is used? Using `__call__` looks correct to me.

Comment (Cyrilvallez) [reply src/transformers/generation/utils.py:3227]: Should be solved with https://github.com/huggingface/transformers/pull/34923 🤗
